File: ios/Approach/Sources/StatisticsLibrary/Protocols/FirstRollStatistic.swift
import Charts
import ModelsLibrary
import StringsLibrary

public protocol FirstRollStatistic: PercentageStatistic, TrackablePerFirstRoll {
	var totalRolls: Int { get set }

	mutating func tracks(firstRoll: Frame.OrderedRoll, configuration: TrackablePerFrameConfiguration) -> Bool
}

extension FirstRollStatistic {
	public static var numeratorTitle: String { title }
	public static var denominatorTitle: String { Strings.Statistics.Title.totalRolls }
	public static var includeNumeratorInFormattedValue: Bool { true }

	public var denominator: Int {
		get { totalRolls }
		set { totalRolls = newValue }
	}

	public mutating func adjust(byFirstRoll: Frame.OrderedRoll, configuration: TrackablePerFrameConfiguration) {
		totalRolls += 1
		if tracks(firstRoll: byFirstRoll, configuration: configuration) {
			numerator += 1
		}
	}
}

File: android/core/statistics/src/main/java/ca/josephroque/bowlingcompanion/core/statistics/interfaces/FirstRollStatistic.kt
package ca.josephroque.bowlingcompanion.core.statistics.interfaces

import ca.josephroque.bowlingcompanion.core.model.TrackableFrame
import ca.josephroque.bowlingcompanion.core.statistics.R
import ca.josephroque.bowlingcompanion.core.statistics.TrackablePerFirstRoll
import ca.josephroque.bowlingcompanion.core.statistics.TrackablePerFrameConfiguration

interface FirstRollStatistic : PercentageStatistic, TrackablePerFirstRoll {
	var totalRolls: Int

	fun tracksRoll(firstRoll: TrackableFrame.Roll, configuration: TrackablePerFrameConfiguration): Boolean

	override val numeratorTitleResourceId: Int
		get() = id.titleResourceId

	override val denominatorTitleResourceId: Int
		get() = R.string.statistic_title_total_rolls

	override val includeNumeratorInFormattedValue: Boolean
		get() = true

	override var denominator: Int
		get() = totalRolls
		set(value) {
			totalRolls = value
		}

	override fun adjustByFirstRoll(firstRoll: TrackableFrame.Roll, configuration: TrackablePerFrameConfiguration) {
		totalRolls++
		if (tracksRoll(firstRoll, configuration)) {
			numerator++
		}
	}
}
